decisions/DR002-Versioning-traits-and-specifications-codegen.md

```diff
@@ -0,0 +1,231 @@
+  requiring use of qualified names (e.g. `v1.traits.LocatableContent`)
+  or additional aliasing (e.g. 
+ `from ... import LocatableContent as LocatableContent_v1`).
+- Not clear at-a-glance which traits have changed between schema
+  versions, e.g. it's not clear if
+  `v2.traits.content.LocatableContentTrait` is the same as
+  `v1.traits.content.LocatableContentTrait`.
+- Must compare multiple YAML documents side-by-side in order to discover
+  the history of changes to a particular trait/specification.
+- Traits/specifications that are unchanged between versions implies
+  duplicated code across namespaces (though likely simply aliased).
+- Independently generated/installed subpackages for each schema version
+  would mean that deprecation warnings could not be added to old
+  versions. This is mitigated if multiple versions are generated
+  together, where the older version can be detected and deprecation
+  warnings added by codegen.
+
+### Option 2 - Per Trait/Specification versioning
+
+A single YAML document is maintained, where each trait/specification
+definition branches off into a list of versions. Old
+trait/specification versions can be marked as deprecated and removed
+eventually, to prevent infinite growth.
+
+For example
+
+```python
+from openassetio_mediacreation.traits.content import LocatableContent_v1
+from openassetio_mediacreation.traits.content import LocatableContent_v2
+from openassetio_mediacreation.specifications.twoDimensional import ImageSpecification_v2
+```
+
+#### Pros
+
+- Fairly trivial to say that the first version "`_v1`" is equivalent to
+  "" (blank), and to ensure that v1's trait ID doesn't contain a version
+  tag, then e.g. the `LocatableContent` class continues to work as
+  before versioning was introduced, making this option fully source
+  compatible with legacy code. I.e. not a breaking change.
+- Placing versions alongside one-another in the YAML definition allows
+  easy discovery of the history of changes.
+- IDE code completion will list all versions of a Trait/Specification
+  view class next to one-another.
+
+#### Cons
+
+- No indication of the version of the constituent traits from the
+  version of a Specification view class.
+- Large change to `traitgen` tool and non-trivial breaking change to
+  YAML documents.
+- Keeping old versions in a living document (as opposed to e.g. git
+  history) is a potential source of accidental breakages to backward
+  compatibility.
+- Generating all possible versions bloats an application's distribution,
+  when it may only use a small subset of them.
+- Higher level branching on a schema version is never possible.
+- A specification's version must be bumped when a constituent trait has
+  a version bump, even if nothing else in the specification has changed.
+  Conceptually, specifications are trait version agnostic, but must
+  become version-aware for the purposes of codegen, which is
+  inconsistent.
```

Comment:
[comment] This really sucks dosent it? Such an obvious entry point for bugs in the actual data. I wonder if there's a way we can guard against this.

Reply:
Would be great to scheme on mitigations. It was my main complaint I had about the move to per-trait versioning in the design sketch.

Hopefully any mitigations we come up with can be added as a non-breaking change (e.g. utility functions) at a later date, rather than dragging out this decision?

Reply:
Yeah, I also hope so, and think they would as I suspect they would be more generation time checks. Definately best to push on in the name of progress.